### PR TITLE
Implement promote/demote feature

### PR DESCRIPTION
**Pull request by princemaple**
#544 We probably need a follow up to add super admin or site owner.

image

image

image

image

### CODE REVIEW

File: apps/fz_http/lib/fz_http_web/live/user_live/show_live.ex [hunk 1]
```diff
@@ -63,4 +63,56 @@ defmodule FzHttpWeb.UserLive.Show do
...
+    "promote"
+  end
+
+  def mote_message(%{role: :admin}) do
+    "This will remove admin permissions from the user."
+  end
+
+  def mote_message(%{role: _}) do
+    "This will give admin permissions to the user."
+  end
```
Comment (princemaple) [inline apps/fz_http/lib/fz_http_web/live/user_live/show_live.ex]: Need help with a better name :)

Also, are they supposed to be here? Or is there a better place in live view structure?

Comment (jamilbk) [reply apps/fz_http/lib/fz_http_web/live/user_live/show_live.ex]: This looks good for now. Should these be private functions, or were you planning to unit test them?
Also, should we make these messages module attributes?

**Comment by jamilbk** [comment]
@princemaple Nice work. I'll have time later today to take a more in-depth look but seems good so far.

Re: the site admin, we allow users to create or reset admins with `firezone-ctl create-or-reset-admin` -- this only works for the user with the admin email found in the config file. Perhaps we can start naming this the site admin... will have to think more about RBAC.

File: apps/fz_http/lib/fz_http_web/live/user_live/show_live.ex (hunk 1 above)
Comment (princemaple) [reply apps/fz_http/lib/fz_http_web/live/user_live/show_live.ex]: Refactored this part. Somehow I was under the impression that private functions are not accessible in the template.

**Comment by princemaple** [comment]
Force pushed with all commits signed

File: apps/fz_http/lib/fz_http_web/live/user_live/show_live.ex [hunk 2]
```diff
@@ -63,4 +63,59 @@ defmodule FzHttpWeb.UserLive.Show do
...
+
+      role =
+        case action do
+          "promote" -> :admin
+          "demote" -> :unprivileged
+        end
+
+      case Users.update_user(user, %{role: role}) do
+        {:ok, user} ->
+          # Force reconnect with new role
```
Comment (jamilbk) [inline apps/fz_http/lib/fz_http_web/live/user_live/show_live.ex:84]: Good catch.

File: apps/fz_http/lib/fz_http_web/live/user_live/show_live.ex [hunk 3]
```diff
@@ -63,4 +63,59 @@ defmodule FzHttpWeb.UserLive.Show do
...
+           socket
+           |> put_flash(
+             :error,
+             "Error updating user: #{ErrorHelpers.aggregated_errors(changeset)}"
+           )}
+      end
+    end
+  end
+
+  @action_and_message %{
```
Comment (jamilbk) [inline apps/fz_http/lib/fz_http_web/live/user_live/show_live.ex:103]: nit: Thoughts on keeping module attributes at the top of the file?

File: apps/fz_http/test/fz_http_web/live/user_live/show_test.exs [hunk 4]
```diff
@@ -479,6 +479,53 @@ defmodule FzHttpWeb.UserLive.ShowTest do
     end
   end
 
+  describe "user role" do
+    setup do
+      admin_user = FzHttp.UsersFixtures.user(role: :admin)
```
Comment (jamilbk) [inline apps/fz_http/test/fz_http_web/live/user_live/show_test.exs]: nit: Thoughts on aliasing `FzHttp.UsersFixtures` at the top of the file?

File: apps/fz_http/test/support/fixtures/users_fixtures.ex [hunk 5]
```diff
@@ -10,18 +10,18 @@ defmodule FzHttp.UsersFixtures do
...
   def user(attrs \\ %{}) do
-    email = Map.get(attrs, :email, "test-#{counter()}@test")
+    email = attrs[:email] || "test-#{counter()}@test"
 
     case Repo.get_by(User, email: email) do
       nil ->
         {:ok, user} =
-          %{
+          attrs
+          |> Enum.into(%{
```
Comment (jamilbk) [inline apps/fz_http/test/support/fixtures/users_fixtures.ex:19]: 👍🏼

**Review by jamilbk — COMMENTED** [review]
Looks good! Just had a couple questions concerning style. I haven't worked on tons of Elixir projects with others, so I'm curious to hear your input.

**Comment by jamilbk** [comment]
@princemaple I'm trying to stick to some sort of conventions around branch naming and commit message structure to make them easier to refer to from the [releases](https://github.com/firezone/firezone/releases/) pages, and make the changelogs more automatable.

Could you rename this feature branch to `{issue_num}/{issue_title}`? E.g. `544/promote_demote_user_role`

Will probably be adopting [conventional commits](https://conventionalcommits.org) in the near future.

**Comment by princemaple** [comment]
What I do with Swoosh is I use [Release Drafter](https://github.com/swoosh/swoosh/blob/main/.github/release-drafter.yml) to automatically generate release notes with tags. The good thing about it is not only that it's automatic, but also when you have to fix wrong tags, it can simply regenerate the whole release notes, all without asking the contributor to do anything. What do you think?

That said, I'm happy to follow a convention if we decide to put one in place.

File: apps/fz_http/test/fz_http_web/live/user_live/show_test.exs (hunk 4 above)
Comment (princemaple) [reply apps/fz_http/test/fz_http_web/live/user_live/show_test.exs]: Sure

File: apps/fz_http/lib/fz_http_web/live/user_live/show_live.ex (hunk 3 above)
Comment (princemaple) [reply apps/fz_http/lib/fz_http_web/live/user_live/show_live.ex:103]: For things that aren't affecting the whole module, and are kind of implementation details, I'd normally keep them where they are used. one example would be https://cs.github.com/elixir-ecto/ecto/blob/0d82133d4d643dcf7007e20dceafefae88453ab9/lib/ecto/query.ex#L1053

Comment (jamilbk) [reply apps/fz_http/lib/fz_http_web/live/user_live/show_live.ex:103]: Kk

**Comment by jamilbk** [comment]
> What I do with Swoosh is I use [Release Drafter](https://github.com/swoosh/swoosh/blob/main/.github/release-drafter.yml) to automatically generate release notes with tags. The good thing about it is not only that it's automatic, but also when you have to fix wrong tags, it can simply regenerate the whole release notes, all without asking the contributor to do anything. What do you think?
> 
> That said, I'm happy to follow a convention if we decide to put one in place.

Awesome — release drafter looks great. I trust your judgement here.

**Review by jamilbk — APPROVED** [review]

**Pull request merged**